File: src/bin/0239-sliding-window-maximum.cpp
// @leetup=info id=239 lang=cpp slug=sliding-window-maximum

#include <iostream>

// @leetup=code

#include <deque>
#include <vector>

class Solution {
 public:
  std::vector<int> maxSlidingWindow(std::vector<int>& nums, int k) {
    std::deque<int> q;
    std::vector<int> res;
    for (int i = 0; i < nums.size(); ++i) {
      int n = nums[i];
      while (!q.empty() && nums[q.back()] <= n) {
        q.pop_back();
      }
      q.push_back(i);
      if (q.front() < i - (k - 1)) {
        q.pop_front();
      }
      if (i >= k - 1) {
        res.push_back(nums[q.front()]);
      }
    }
    return res;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0239-sliding-window-maximum.py
# @leetup=info id=239 lang=python3 slug=sliding-window-maximum



# @leetup=code

from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        N = len(nums)
        # q stores index
        q = deque()
        res = []
        for i in range(N):
            n = nums[i]
            # Make q monotonic decreasing queue.
            while q and nums[q[-1]] <= n:
                q.pop()
            q.append(i)

            if q[0] < i - (k - 1):
                # q[0] is out of sliding window.
                q.popleft()

            if i >= k - 1:
                res.append(nums[q[0]])

        return res


# @leetup=code
